uniswap_edpr/src/logics/shared_algorithm_lens/src/lib.rs

```diff
@@ -0,0 +1,86 @@
+use shared_algorithm_lens_accessors :: * ;
+
+# [derive (Clone , Debug , Default , candid :: CandidType , serde :: Deserialize , serde :: Serialize)]
+pub struct LensValue {
+  pub address: String,
+  pub current_price: f32,
+  pub range_top: f32,
+  pub range_bottom: f32,
+  pub edpr: f32
+}
+
+pub async fn calculate (targets : Vec < String >) -> LensValue {
+  let pool_fees_result = get_get_last_snapshot_value_in_pool_fees (targets . get (0usize) . unwrap () . clone ()) . await ;
+  let tc28x6_result = get_get_last_snapshot_value_in_tcumul_28x6hr (targets . get (1usize) . unwrap () . clone ()) . await ;
+  let v3pool_result = get_get_last_snapshot_value_in_v3pool (targets . get (2usize) . unwrap () . clone ()) . await ;
+  let eth_usdc_price_result = get_get_last_snapshot_value_in_eth_usdc_price (targets . get (3usize) . unwrap () . clone ()) . await ;
+  
+  let fees_24h_usd = pool_fees_result.clone().unwrap().result[0].fees_24h_usd;
+  let t0_decimals = pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t0_decimals;
+  let t1_decimals =  pool_fees_result.clone().unwrap().result[0].pool_summary_level_1.t1_decimals;
+
+  let eth_price_usd  = eth_usdc_price_result.clone().unwrap().result.t1_price_usd;
+
+  let address = v3pool_result.clone().unwrap().result.address;
+  let current_tick_liquidity_str = v3pool_result.clone().unwrap().result.liquidity;
+  let current_tick_liquidity = i128::from_str_radix(current_tick_liquidity_str.trim_start_matches("0x"), 16).unwrap();
+  let sqrt_ratio_x96_str = v3pool_result.clone().unwrap().result.sqrt_ratio_x96;
+  let sqrt_ratio_x96 = u128::from_str_radix(sqrt_ratio_x96_str.trim_start_matches("0x"), 16).unwrap();
+  let tick_current = v3pool_result.clone().unwrap().result.tick_current;
+  let tick_spacing = v3pool_result.clone().unwrap().result.tick_spacing;
+  let token0 = v3pool_result.clone().unwrap().result.token0;
+
+  let tick_cumul_28x6h = tc28x6_result.unwrap().0;
+
+  let current_sqrt_price = sqrt_ratio_x96 as f32 / f32::powf(2.0,96.0);
+  let mut current_price = current_sqrt_price * current_sqrt_price;
```

Comment:
Just to be sure, this is always the dollar equivalent (not the relative value in the pool), right?

Reply:
No, it's the relative value in the pool. 

The same applies to range_top and range_bottom. This is the more helpful way to view it when it comes to [providing liquidity](https://app.uniswap.org/add/0xd33526068d116ce69f19a9ee46f0bd304f21a51f/0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2/3000?minPrice=0.012134394&maxPrice=0.014449563).

LensValue { 
  address: "0xe42318ea3b998e8355a3da364eb9d48ec725eb45", 
  current_price: 0.013549291, 
  range_top: 0.014449563, 
  range_bottom: 0.012134394, 
  edpr: 0.04053104 
}

![image](https://github.com/horizonx-tech/chainsight-showcase/assets/37236322/4c36c223-b1f8-40ca-a9a0-a087f689a7cb)

Reply:
I see, I guess the dollar conversion is generally easier to understand, but it has that advantage.
Let's leave it as it is this time. But it still seems to me that at least the following pointed comment is necessary.
https://github.com/horizonx-tech/chainsight-showcase/pull/34#discussion_r1450532067

Reply:
Agreed. Thanks.